folding-schemes/src/folding/nova/nifs/mova.rs

```diff
@@ -0,0 +1,415 @@
+            rE = (0..log2(W.E.len()))
+                .map(|_| C::ScalarField::rand(&mut rng))
+                .collect();
+        }
+
+        W.commit::<CS, H>(params, x, rE)
+    }
+
+    // Protocol 7 - point 3 (16)
+    fn fold_witness(
+        a: C::ScalarField,
+        W_i: &Witness<C>,
+        w_i: &Witness<C>,
+        aux: &Vec<C::ScalarField>, // T in Mova's notation
+    ) -> Result<Witness<C>, Error> {
+        let a2 = a * a;
+        let E: Vec<C::ScalarField> = vec_add(
+            &vec_add(&W_i.E, &vec_scalar_mul(aux, &a))?,
+            &vec_scalar_mul(&w_i.E, &a2),
+        )?;
+        let W: Vec<C::ScalarField> = W_i
+            .W
+            .iter()
+            .zip(&w_i.W)
+            .map(|(i1, i2)| *i1 + (a * i2))
+            .collect();
+
+        let rW = W_i.rW + a * w_i.rW;
+        Ok(Witness::<C> { E, W, rW })
+    }
+
+    /// [Mova](https://eprint.iacr.org/2024/1220.pdf)'s section 4. Protocol 8
+    /// Returns a proof for the pt-vs-line operations along with the folded committed instance
+    /// instances and witness
+    #[allow(clippy::type_complexity)]
+    fn prove(
+        _cs_prover_params: &CS::ProverParams, // not used in Mova since we don't commit to T
+        r1cs: &R1CS<C::ScalarField>,
+        transcript: &mut T,
+        pp_hash: C::ScalarField,
+        W_i: &Witness<C>,
+        U_i: &CommittedInstance<C>,
+        w_i: &Witness<C>,
+        u_i: &CommittedInstance<C>,
+    ) -> Result<
+        (
+            Self::Witness,
+            Self::CommittedInstance,
+            Self::Proof,
+            Vec<bool>,
+        ),
+        Error,
+    > {
+        transcript.absorb(&pp_hash);
+        // Protocol 5 is pre-processing
+        transcript.absorb(U_i);
+        transcript.absorb(u_i);
+
+        // Protocol 6
+        let (
+            h_proof,
+            PointvsLineEvaluationClaim {
+                mleE1_prime,
+                mleE2_prime,
+                rE_prime,
+            },
+        ) = PointVsLine::<C, T>::prove(transcript, U_i, u_i, W_i, w_i)?;
+
+        // Protocol 7
+
+        transcript.absorb(&mleE1_prime);
+        transcript.absorb(&mleE2_prime);
+
+        // compute the cross terms
+        let z1: Vec<C::ScalarField> = [vec![U_i.u], U_i.x.to_vec(), W_i.W.to_vec()].concat();
+        let z2: Vec<C::ScalarField> = [vec![u_i.u], u_i.x.to_vec(), w_i.W.to_vec()].concat();
+        let T = NovaNIFS::<C, CS, T, H>::compute_T(r1cs, U_i.u, u_i.u, &z1, &z2)?;
+
+        let n_vars: usize = log2(W_i.E.len()) as usize;
+        if log2(T.len()) as usize != n_vars {
+            return Err(Error::NotExpectedLength(T.len(), n_vars));
+        }
+
+        let mleT = dense_vec_to_dense_mle(n_vars, &T);
+        let mleT_evaluated = mleT.evaluate(&rE_prime).ok_or(Error::EvaluationFail)?;
+
+        transcript.absorb(&mleT_evaluated);
+
+        let alpha_scalar = C::ScalarField::from_le_bytes_mod_order(b"alpha");
+        transcript.absorb(&alpha_scalar);
```

Comment:
We dont actually need the alpha_scalar. I removed it in my latest commit

Reply:
oh right, thanks, [removed](https://github.com/NiDimi/sonobe/compare/8e57b85bdba156fef13028f6e028879b90529af4..54e6b60a60115b40d31ca0a9fff2fbccbdb47ea9)